Drive to point using heuristics

JohnGilb opened the pull request:
This is a simple set of heuristics that attempts to reach a final position/heading from an initial position/heading by:
1) Driving towards the line described by extending the heading from the final position
2) Orient towards the final orientation
3) Drive towards the final point

I recommend looking at https://github.com/Team488/SeriouslyCommonLib/pull/81 for context on how the math is done.

JohnGilb committed:
start the DriveToPointUsingHeuristics

JohnGilb committed:
Complete one test case, fixing bugs along the way

JohnGilb committed:
Make the effect distances a property

JohnGilb committed:
remove unused import

WasabiFan commented on Competition2017/src/competition/subsystems/drive/commands/DriveToPointUsingHeuristicsCommand.java:
```diff
@@ -0,0 +1,120 @@
+        super(driveSubsystem);
+        this.poseSubsystem = poseSubsystem;
+        
+        alignToFinalHeadingPid = pidFactory.createPIDManager("AlignToFinalHeading", .1, 0, 0, 0, 1, -1, 2, 2, 1);
```
Spacing, caps

JohnGilb replied:
Cleaned up some of these

WasabiFan commented on Competition2017/src/competition/subsystems/drive/commands/DriveToPointUsingHeuristicsCommand.java:
```diff
@@ -0,0 +1,120 @@
+        double heading = currentPoint.getHeading().getValue();
+        
+        // x and y to add
+        double yDelta = (Math.sin(Math.toRadians(heading)) * 60) + (Math.sin(Math.toRadians(heading+90)) * 60);
```
What's `60`? Should that be parameterized (ref line 48)?

JohnGilb replied:
parameterized!

WasabiFan commented on Competition2017/src/competition/subsystems/drive/commands/DriveToPointUsingHeuristicsCommand.java:
```diff
@@ -0,0 +1,120 @@
+                new XYPair(
+                        currentPoint.getPoint().x + xDelta,
+                        currentPoint.getPoint().y + yDelta),
+                new ContiguousHeading(heading + 90));
```
Same as above

JohnGilb replied:
parameterized!

WasabiFan commented on Competition2017/src/competition/subsystems/drive/commands/DriveToPointUsingHeuristicsCommand.java:
```diff
@@ -0,0 +1,120 @@
+        
+        ContiguousHeading alignTowardsLineHeading = 
+                goalPoint.getLinePerpendicularToPoint(currentPose.getPoint()).getHeading();
+        alignTowardsLineHeading = yDisplacement > 0 ? alignTowardsLineHeading : alignTowardsLineHeading.shiftValue(180);
```
Does this properly function when `yDisplacement == 0`?

JohnGilb replied:
I think it's mostly a moot point - if `yDisplacement` is at zero, it means (at worst) that you're oscillating between two opposite points on a circle, which are essentially equivalent in terms of how they will make the robot turn to align towards the line.

WasabiFan commented on Competition2017/src/competition/subsystems/drive/commands/DriveToPointUsingHeuristicsCommand.java:
```diff
@@ -0,0 +1,120 @@
+    
+    @Override
+    public boolean isFinished() {
+        boolean reachedDestination = destinationPid.isOnTarget();
```
I think this would be clearer if the `isOnTarget` calls were inlined

JohnGilb replied:
good call.

WasabiFan commented on Competition2017/tests/competition/subsystems/drive/DriveTestBase.java:
```diff
@@ -53,10 +53,60 @@ public void verifyDriveNegative() {
+        pose.updatePeriodicData();
     }
     
     public void driveDeltaEncoderDistances(double left, double right){
```
Is this in ticks? Now that there's an `inches` version, I think this one should have the units in the name too.

JohnGilb replied:
Seems ok

WasabiFan commented on Competition2017/tests/competition/subsystems/drive/DriveTestBase.java:
```diff
@@ -53,10 +53,60 @@ public void verifyDriveNegative() {
+        assertTrue((getLeftSetpoint() - getRightSetpoint()) > minimumDifference);
+    }
+    
+    public void verifyDriveArcingRight() {
```
Missing corresponding function for left side

JohnGilb replied:
Fixed - logic was split across several functions =[

WasabiFan commented on Competition2017/tests/competition/subsystems/drive/commands/DriveToPointUsingHeuristicsCommandTest.java:
```diff
@@ -0,0 +1,79 @@
+    
+    @Before
+    public void setUp() {
+        // TODO Auto-generated method stub
```
Remove TODOs

JohnGilb replied:
Good catch

WasabiFan reviewed (commented):
I didn't go through the logic in great detail, but it looks generally good. It might be helpful to add some more rigorous logging.

aschokking commented on Competition2017/src/competition/subsystems/drive/commands/DriveForDistanceCommand.java:
```diff
@@ -24,6 +24,8 @@
     private DoubleProperty deltaDistanceProp;
     private ContiguousHeading targetHeading;
     
+    public static String driveToPositionPidName = "Drive to position";
```
shouldn't this then get referenced in the pidmanager construction? Otherwise they could get out of sync?
while we're here it's a little strange that this name doesn't match either the class or the variable name in code of the property, wonder if there's some cheap rationalize we could do there?

JohnGilb replied:
Orphaned code from an earlier implementation, removed.

aschokking commented on Competition2017/src/competition/subsystems/drive/commands/DriveToPointUsingHeuristicsCommand.java:
```diff
@@ -0,0 +1,120 @@
+        super(driveSubsystem);
+        this.poseSubsystem = poseSubsystem;
+        
+        alignToFinalHeadingPid = pidFactory.createPIDManager("AlignToFinalHeading", .1, 0, 0, 0, 1, -1, 2, 2, 1);
```
consider maybe 'targetFinalHeading' instead, I think we use that more often to describe the setpoint of such a system?

WasabiFan replied:
This isn't a setpoint; it's a PID manager name.

aschokking replied:
Ah you're right my bad.

aschokking commented on Competition2017/src/competition/subsystems/drive/commands/DriveToPointUsingHeuristicsCommand.java:
```diff
@@ -0,0 +1,120 @@
+        
+        alignToFinalHeadingPid = pidFactory.createPIDManager("AlignToFinalHeading", .1, 0, 0, 0, 1, -1, 2, 2, 1);
+        headTowardsLinePid = pidFactory.createPIDManager("HeadTowardsLine", .1, 0, 0, 0, 1, -1, 2, 2, 1);
+        destinationPid = pidFactory.createPIDManager("DriveToDestination", 0.1, 0, 0, 0, 1, -1, 3, 1, 0.5);
```
It might be hard to tell the difference between this and the drivefordistance's pid properties based on just the name?

JohnGilb replied:
added a better prefix

aschokking commented on Competition2017/src/competition/subsystems/drive/commands/DriveToPointUsingHeuristicsCommand.java:
```diff
@@ -0,0 +1,120 @@
+        double yDisplacement = goalPoint.getPointRelativeYDisplacementFromLine(currentPose);
+        
+        // once we are within 6 inches, this begins to get important
+        double finalHeadingScalingFactor = finalHeadingEffectDistance.get() / (12 + finalHeadingEffectDistance.get());
```
what's the 12?

JohnGilb replied:
code bug! fixed.

aschokking commented on Competition2017/tests/competition/subsystems/drive/DriveTestBase.java:
```diff
@@ -53,10 +53,60 @@ public void verifyDriveNegative() {
+    }
+    
+    public void verifyNotTurning(double minimumDifference) {
+        assertTrue((getLeftSetpoint() - getRightSetpoint()) < minimumDifference);
```
curious why not use assertEquals with the diff as the 3rd error arg?

aschokking replied:
Using assertEquals also has some nice advantages like when this fails it'll print out what the values were as opposed to just saying an assertion failed

JohnGilb replied:
good thought!

aschokking commented on Competition2017/tests/competition/subsystems/drive/DriveTestBase.java:
```diff
@@ -53,10 +53,60 @@ public void verifyDriveNegative() {
+    }
+    
+    public void verifyStopped(double minimumPower) {
+        assertTrue(Math.abs(getLeftSetpoint()) < minimumPower);
```
same thing here, consider assertEquals(, 0, minimumPower)

JohnGilb replied:
yep

aschokking commented on Competition2017/tests/competition/subsystems/drive/commands/DriveToPointUsingHeuristicsCommandTest.java:
```diff
@@ -0,0 +1,79 @@
+    }
+    
+    @Test
+    public void start() {
```
it might be nice to give a little more descriptive test method name about what this scenario is

JohnGilb replied:
updated

aschokking reviewed (approved):
I'm excited to see this in practice and have you draw some lines out on the white board about it :)

JohnGilb committed:
point to newer commonlib.

-Fix bug in DriveToPointUsingHeuristicsCommand
-Add ability to set an absolute position goal, or a delta position goal

JohnGilb committed:
Avoid double-calling setDeltaBasedTravel

JohnGilb committed:
PR Feedback - fixing several bugs, increasing name clarity

JohnGilb committed:
Merged 'master'.

WasabiFan commented on Competition2017/src/competition/subsystems/drive/commands/DriveToPointUsingHeuristicsCommand.java:
```diff
@@ -0,0 +1,164 @@
+        if (goalPoint == null) {
+            // If no goal point was set, this should just return immediately.
+            log.warn("DriveToPointUsingHeuristics was started, but no goal point was set. "
+                    + "Command can not continue.");
```
`can not` -> `cannot`; "can not" is the opposite of "cannot" :wink:

JohnGilb committed:
Add DriveToPointUsingHeuristicsCommand to the dashboard

JohnGilb committed:
small spelling updates =]
